zrml/swaps/src/weights.rs

```diff
@@ -288,4 +293,63 @@ impl<T: frame_system::Config> WeightInfoZeitgeist for WeightInfo<T> {
             .saturating_add(T::DbWeight::get().writes(1 as Weight))
             .saturating_add(T::DbWeight::get().writes((2 as Weight).saturating_mul(a as Weight)))
     }
+    // Storage: Swaps PoolsCachedForArbitrage (r:1 w:0)
+    // Storage: Swaps Pools (r:6 w:0)
+    // Storage: Tokens Accounts (r:390 w:390)
+    // Storage: Tokens TotalIssuance (r:65 w:65)
+    fn apply_to_cached_pools_execute_arbitrage(a: u32) -> Weight {
+        (135_773_000 as Weight)
+            // Standard Error: 35_228_000
+            .saturating_add((3_291_495_000 as Weight).saturating_mul(a as Weight))
+            .saturating_add(T::DbWeight::get().reads(45 as Weight))
+            .saturating_add(T::DbWeight::get().reads((67 as Weight).saturating_mul(a as Weight)))
+            .saturating_add(T::DbWeight::get().writes(44 as Weight))
+            .saturating_add(T::DbWeight::get().writes((66 as Weight).saturating_mul(a as Weight)))
+    }
```

Comment:
> Wow, heavy heavy and it depends on `a` (number of cached pools and I fear, that they are really common, because they are cached with every pool swap action AFAIU). 😳

But they are only cached at most once per block. Also, compare `execute_arbitrage_mint_sell(asset_count, 1)` to `pool_join(asset_count)`. Basically twice as heavy, so `execute_arbitrage_mint_sell(a, 1)` is `O(pool_join(a))`.

> How would the weights be changed, when you use the number of assets as a benchmark component? I could imagine, that this makes a huge difference for the overall weight function.

Sure, you can see that in the benchmarks of `execute_arbitrage_mint_sell`, etc. which are parametrzied by the number of assets.

> I think all these reads come from the for loop iteration over `MaxAssets` to deposit or withdraw outcome tokens, right?

Yes. Maybe we can shortcut here by reading/writing only once using the prefix? Don't know if `orml_currencies` exposes that abstraction.

Reply:
What do you think about executing the on-chain arbitrage every `% ARBITRAGE_CYCLE` amount of blocks inside the `on_initialize` to guarantee, that it executes a specific number of pools (don't need to be all possible pools)? When I think about `on_idle` I know it executes before `on_finalize`, so the weight of this arbitrage could very often be too big, when I look at the weight functions. But you already figured out, that we are not running at all into weight problems with the current production chain.

Reply:
> What do you think about executing the on-chain arbitrage every `% ARBITRAGE_CYCLE` amount of blocks inside the `on_initialize` to guarantee, that it executes a specific number of pools (don't need to be all possible pools)? When I think about `on_idle` I know it executes before `on_finalize`, so the weight of full arbitrage could very often be too big, when I look at the weight functions. But you already figured out, that we are not running at all into weight problems with the current production chain.

The problem with using `on_initialize` is that the prices aren't corrected by the end of the block. That's confusing. I'd rather avoid this. Of course this can be done, but before anyone overreacts, I'd rather get some numbers on the table.

That being said, now that I look at the benchmarks again, I don't understand why the parameter `b` (the number of iterations) is a factor in the number of database reads occuring. I'll take a look.

Reply:
[d8c913f](https://github.com/zeitgeistpm/zeitgeist/pull/833/commits/d8c913fa5a812431d8bc3c9d07c218fedfe7f316)

Reply:
No idea why the benchmarks messed that up. I guess it's some kind of a quadratic dependency (each iteration takes a little longer depending on the number of assets) that confused the benchmarks. We now just benchmark the worst-case in terms of iterations. That's fine by me, considering how few iterations we perform.

Reply:
I mean, we should add at least the benchmark component `assets_count` for `apply_to_cached_pools_execute_arbitrage`, so that the weight is a little bit more accurate. Don't you agree? I mean not many markets have got `MaxAsset` assets.

Reply:
Oh, how I wish it was this easy...

First of all, note that we only use these benchmarks when estimating how many pools we can arbitrage. After we're done, we've got a clearer estimate. The advantage of doing it this way is that determining how many pools you can arbitrage is O(1).

The problem with parametrizing the benchmark by the number of assets means that we need to read the number of assets from the chain. So now if you want to arbitrage n pools, even if you cache their asset count, you have to read linear amounts of storage. So now the function which determines if you have enough time to arbitrage all pools is O(n), which means you need a function which determines if you can even run the function which determines if you can run the on-chain arbitrage (yes, that sentence is correct).

For a permanent solution (or if we notice people spamming blocks to prevent arbitrage), I would actually do this. But since this is only a temporary solution...